### PR TITLE
[SPARK-14410] [SQL] : SessionCatalog needs to check database/table/function existence

### PR DESCRIPTION
**Pull request by rekhajoshm**
## What changes were proposed in this pull request?

1.Check for existence of Function in catalog, instead of try-catch
2. Check for existence of database before create/alter/drop database, though ignoreIfExists/ignoreIfNotExists become futile
3.Rename getFunction to getMetadataOfFunction
## How was this patch tested?

TestSuite

### CODE REVIEW

**Comment by yhuai** [comment]
@rekhajoshm Thank you for working on this! I think there are other methods in SessionCatalog that handle an existing table/db/function that do not check the existence of the table/db/function (e.g. dropFunction). Can you go through methods provided by SessionCatalog and make changes accordingly?

**Comment by rekhajoshm** [comment]
@yhuai - will do, one question is., as database calls had ignoreIfExists/ignoreIfNotExists flag, now with adding existence check this boolean loses meaning.Do we still pass that flag? If not, more changes across in clients.Please let me know.thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -84,15 +84,21 @@ class SessionCatalog(
   // ----------------------------------------------------------------------------
 
   def createDatabase(dbDefinition: CatalogDatabase, ignoreIfExists: Boolean): Unit = {
-    externalCatalog.createDatabase(dbDefinition, ignoreIfExists)
+    if (!databaseExists(dbDefinition.name)) {
+      externalCatalog.createDatabase(dbDefinition, ignoreIfExists)
+    }
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: For createDatabase, we should throw an exception if db already exists and ignoreIfExists is false, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 2]
```diff
@@ -84,15 +84,21 @@ class SessionCatalog(
...
   }
 
   def dropDatabase(db: String, ignoreIfNotExists: Boolean, cascade: Boolean): Unit = {
-    externalCatalog.dropDatabase(db, ignoreIfNotExists, cascade)
+    if (databaseExists(db)) {
+      externalCatalog.dropDatabase(db, ignoreIfNotExists, cascade)
+    }
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: This is now wrong. We used to pass the flag `ignoreIfNotExists` to the external catalog, which will handle the exception throwing. Now we'll ignore this flag if the database doesn't exist.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -474,15 +479,8 @@ class SessionCatalog(
...
-      try {
-        // TODO: It's better to ask external catalog if this function exists.
-        // So, we can avoid of having this hacky try/catch block.
-        getFunction(name) != null
-      } catch {
-        case _: NoSuchFunctionException => false
-        case _: AnalysisException => false // HiveExternalCatalog wraps all exceptions with it.
-      }
+      val db = name.database.getOrElse(currentDb)
+      externalCatalog.listFunctions(db,name.unquotedString).size > 0
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: No, the right thing to do here is to implement `functionExists` in the external catalog and `HiveClientImpl`. That's what the TODO meant.

**Comment by andrewor14** [comment]
@rekhajoshm I think there is some misunderstanding on what the issue is. The JIRA is saying all the create/drop methods in `SessionCatalog` should take in a flag `ignoreIf[Not]Exists`. I believe actually create/drop functions are the only places where we need to do that.

**Comment by andrewor14** [comment]
@rekhajoshm I've opened #12198 to illustrate what I mean. I don't think the changes in this patch are correct so I would recommend that we close this PR.

**Comment by rekhajoshm** [comment]
ok thanks @andrewor14

**Pull request closed without merging**